README.md
# deno_brotli

[![stars](https://img.shields.io/github/stars/denosaurs/deno_brotli)](https://github.com/denosaurs/deno_brotli/stargazers)
[![workflow](https://img.shields.io/github/workflow/status/denosaurs/deno_brotli/ci)](https://github.com/denosaurs/deno_brotli/actions)
[![releases](https://img.shields.io/github/v/release/denosaurs/deno_brotli)](https://github.com/denosaurs/deno_brotli/releases/latest/)
[![deno version](https://img.shields.io/badge/deno-^1.0.2-informational)](https://github.com/denoland/deno)
[![deno doc](https://img.shields.io/badge/deno-doc-informational)](https://doc.deno.land/https/deno.land/x/deno_brotli/mod.ts)
[![Discord](https://img.shields.io/discord/713043818806509608)](https://discord.gg/shHG8vg)
[![license](https://img.shields.io/github/license/denosaurs/deno_brotli)](https://github.com/denosaurs/deno_brotli/blob/master/LICENSE)
[![issues](https://img.shields.io/github/issues/denosaurs/deno_brotli)](https://github.com/denosaurs/deno_brotli/issues)

This module provides [brotli](https://en.wikipedia.org/wiki/Brotli) support for
deno and the web by providing [simple bindings](src/lib.rs) using
[rust-brotli](https://github.com/dropbox/rust-brotli) compiled to webassembly.

## Usage

### Compression

```ts
import { compress } from "https://deno.land/x/brotli/mod.ts";

const text = new TextEncoder().encode("X".repeat(64));

console.log(text.length); // 64 Bytes
console.log(compress(text).length); // 10 Bytes
```

### Decompression

```ts
import { decompress } from "https://deno.land/x/brotli/mod.ts";

const compressed = Uint8Array.from([27, 63, 0, 0, 36, 176, 226, 153, 64, 18]);

console.log(compressed.length); // 10 Bytes
console.log(decompress(compressed).length); // 64 Bytes
```

## Other

### Contribution

Pull request, issues and feedback are very welcome. Code style is formatted with
`deno fmt` and commit messages are done following
[Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/) spec.

### Licence

Copyright 2020-2022, the denosaurs team. All rights reserved. MIT license.


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog], and this project adheres to
[Semantic Versioning].

## [0.1.5] - 2020-11-04

### Features

- clean, fmt, lint scripts ([`0283878`])
- benchmarks ([`d20e382`])

## [0.1.4] - 2020-07-14

### Bug Fixes

- :construction_worker: Fix build name ([`4f64f21`])

## [0.1.3] - 2020-06-01

### Bug Fixes

- :fire: Removed wee_alloc to fix memory leak ([`4256de7`])

## [0.1.2] - 2020-05-30

### Features

- :zap: Add settings.json to improve vscode performance ([`a4a3cfb`])

## [0.1.1] - 2020-05-28

### Features

- add js minification to wasm.js ([`0410c54`])

### Bug Fixes

- add opt-level ([`53b18e5`])

## [0.1.0] - 2020-05-28

### Bug Fixes

- :art: deno fmt now ignores wasm.js ([`eade930`])

## [0.0.1] - 2020-05-28

### Features

- add Release and PR CIs ([`a4cc61c`])
- add examples for decompress ([`f670133`])
- add `decompress` to class ([`84a06bc`])
- bind `decompress` to plugin ([`fa5e90d`])
- add `decompress` to rust ([`891bb3d`])
- add `compress` method ([`9432503`])
- initial commit ([`cf3c6c2`])

### Bug Fixes

- decompress fails ([`9165eb1`])

[keep a changelog]: https://keepachangelog.com/en/1.0.0/
[semantic versioning]: https://semver.org/spec/v2.0.0.html
[0.1.5]: https://github.com/denosaurs/deno_brotli/compare/0.1.4...0.1.5
[`0283878`]: https://github.com/denosaurs/deno_brotli/commit/028387867a853ad1a7657b926a09535864400921
[`d20e382`]: https://github.com/denosaurs/deno_brotli/commit/d20e382d7a2780b09887b5a91b51df98adb513ad
[0.1.4]: https://github.com/denosaurs/deno_brotli/compare/0.1.3...0.1.4
[`4f64f21`]: https://github.com/denosaurs/deno_brotli/commit/4f64f21a6b2405117c7f407d0bbbaa25ff9de632
[0.1.3]: https://github.com/denosaurs/deno_brotli/compare/0.1.2...0.1.3
[`4256de7`]: https://github.com/denosaurs/deno_brotli/commit/4256de731322bf5be710169af72cf11333972ff6
[0.1.2]: https://github.com/denosaurs/deno_brotli/compare/0.1.1...0.1.2
[`a4a3cfb`]: https://github.com/denosaurs/deno_brotli/commit/a4a3cfbc2d7edc62ac7fea901e9b4b197d4ab593
[0.1.1]: https://github.com/denosaurs/deno_brotli/compare/0.1.0...0.1.1
[`0410c54`]: https://github.com/denosaurs/deno_brotli/commit/0410c548125f7e793ae0fdaeca4ce4d0cc60928e
[`53b18e5`]: https://github.com/denosaurs/deno_brotli/commit/53b18e5e3994aaebc017403e686a6a7a1249c69a
[0.1.0]: https://github.com/denosaurs/deno_brotli/compare/0.0.1...0.1.0
[`eade930`]: https://github.com/denosaurs/deno_brotli/commit/eade930b76e317d190903e4dbd56bb71e87ffbb8
[0.0.1]: https://github.com/denosaurs/deno_brotli/compare/0.0.1
[`a4cc61c`]: https://github.com/denosaurs/deno_brotli/commit/a4cc61c862dadcd0aae80a02942a320036ea3285
[`f670133`]: https://github.com/denosaurs/deno_brotli/commit/f6701338a4c8b9a4c0f0b93b45d5f6f7cac67e64
[`84a06bc`]: https://github.com/denosaurs/deno_brotli/commit/84a06bcb1eb249de66b1dd9e58fdfc52986780d9
[`fa5e90d`]: https://github.com/denosaurs/deno_brotli/commit/fa5e90d6c990f9be46b3c615a5faf77952620954
[`891bb3d`]: https://github.com/denosaurs/deno_brotli/commit/891bb3d3746e235406b06743f86cfda454506c87
[`9432503`]: https://github.com/denosaurs/deno_brotli/commit/94325038308388807b9d08831fb3830d9bf1b3ff
[`cf3c6c2`]: https://github.com/denosaurs/deno_brotli/commit/cf3c6c2e929a286a510b091c8cab494e79643f0b
[`9165eb1`]: https://github.com/denosaurs/deno_brotli/commit/9165eb1539999a1b5707559d6e6eee0d6ff7f7f9


mod.ts
import init, {
  compress as wasmCompress,
  decompress as wasmDecompress,
  source,
} from "./wasm.js";

await init(source);

/**
 * Compress a byte array.
 *
 * ```typescript
 * import { compress } from "https://deno.land/x/brotli/mod.ts";
 * const text = new TextEncoder().encode("X".repeat(64));
 * console.log(text.length);                   // 64 Bytes
 * console.log(compress(text).length);         // 10 Bytes
 * ```
 *
 * @param input Input data.
 * @param bufferSize Read buffer size
 * @param quality Controls the compression-speed vs compression-
 * density tradeoff. The higher the quality, the slower the compression.
 * @param lgwin Base 2 logarithm of the sliding window size.
 */
export function compress(
  input: Uint8Array,
  bufferSize = 4096,
  quality = 6,
  lgwin = 22,
): Uint8Array {
  return wasmCompress(input, bufferSize, quality, lgwin);
}

/**
 * Decompress a byte array.
 *
 * ```typescript
 * import { decompress } from "https://deno.land/x/brotli/mod.ts";
 * const compressed = Uint8Array.from([ 27, 63, 0, 0, 36, 176, 226, 153, 64, 18 ]);
 * console.log(compressed.length);             // 10 Bytes
 * console.log(decompress(compressed).length); // 64 Bytes
 * ```
 *
 * @param input Input data.
 * @param bufferSize Read buffer size
 */
export function decompress(
  input: Uint8Array,
  bufferSize = 4096,
): Uint8Array {
  return wasmDecompress(input, bufferSize);
}
